File: services/docgen/app/schemas/storage_gc.py
from __future__ import annotations

from pydantic import BaseModel, Field


class StorageGcInventory(BaseModel):
    jobs_with_zip: int = 0
    jobs_eligible: int = 0
    jobs_purged_history: int = 0
    retention_days: int | None = None
    cutoff: str | None = None


class StorageGcStats(BaseModel):
    retention_days: int | None = None
    cutoff: str | None = None
    jobs_with_zip: int | None = None
    jobs_eligible: int | None = None
    jobs_scanned: int = 0
    jobs_purged: int = 0
    objects_deleted: int = 0
    errors: int = 0
    forced: bool = False
    purge_all: bool = False
    skipped: bool = False
    reason: str | None = None
    last_run_at: str | None = None


class StorageGcConfigResponse(BaseModel):
    enabled: bool
    retention_days: int = Field(ge=1, le=365)
    last_run_at: str | None = None
    last_stats: StorageGcStats | None = None
    inventory: StorageGcInventory


class StorageGcConfigUpdate(BaseModel):
    enabled: bool | None = None
    retention_days: int | None = Field(default=None, ge=1, le=365)


class StorageGcRunResponse(BaseModel):
    message: str
    stats: StorageGcStats

File: frontend/src/api/storageGc.ts
import { apiFetch } from "./client";

export interface StorageGcInventory {
  jobs_with_zip: number;
  jobs_eligible: number;
  jobs_purged_history: number;
  retention_days?: number;
  cutoff?: string;
}

export interface StorageGcStats {
  retention_days?: number;
  cutoff?: string;
  jobs_with_zip?: number;
  jobs_eligible?: number;
  jobs_scanned?: number;
  jobs_purged?: number;
  objects_deleted?: number;
  errors?: number;
  forced?: boolean;
  purge_all?: boolean;
  skipped?: boolean;
  reason?: string;
  last_run_at?: string;
}

export interface StorageGcConfig {
  enabled: boolean;
  retention_days: number;
  last_run_at: string | null;
  last_stats: StorageGcStats | null;
  inventory: StorageGcInventory;
}

export interface StorageGcRunResult {
  message: string;
  stats: StorageGcStats;
}

export interface GenerationStatusCount {
  status: string;
  count: number;
}

export interface GenerationMonthCount {
  month: string;
  jobs: number;
  certificates: number;
}

export interface GenerationEventTop {
  event_id: string;
  event_title: string;
  project_number: string;
  jobs: number;
  certificates: number;
}

export interface StorageGcAnalytics {
  total_jobs: number;
  total_certificates: number;
  success_rate: number;
  by_status: GenerationStatusCount[];
  by_month: GenerationMonthCount[];
  top_events: GenerationEventTop[];
  inventory: StorageGcInventory;
}

export function fetchStorageGcAnalytics(token: string | null) {
  return apiFetch<StorageGcAnalytics>("/v1/generations/admin/storage-gc/analytics", token);
}

export function fetchStorageGcConfig(token: string | null) {
  return apiFetch<StorageGcConfig>("/v1/generations/admin/storage-gc/config", token);
}

export function updateStorageGcConfig(
  token: string | null,
  payload: { enabled?: boolean; retention_days?: number },
) {
  return apiFetch<StorageGcConfig>("/v1/generations/admin/storage-gc/config", token, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export function runStorageGc(token: string | null, options?: { purgeAll?: boolean }) {
  const query = options?.purgeAll ? "?purge_all=true" : "";
  return apiFetch<StorageGcRunResult>(`/v1/generations/admin/storage-gc/run${query}`, token, {
    method: "POST",
  });
}
